### ISSUE TITLE
backfill_from does not always deliver most recent data

### ISSUE THREAD

**Issue by rterbush** [mementum/backtrader#269]
I'm still trying to collect info on this but want to open this issue as a place to start collecting evidence.

Some days(not clear if it is after gap days, odd days, etc) the data that is delivered by `backfill_from=` does not show the last bar to a daily timeframe strategy. 

Today, even though the bar is present in the CSV data for 2/17/2017, the data is not seen in `strategy.next()`

Other possible symptom I am seeing is that indicator values are different in live trade mode vs. backtested values.

```
-- 5454 2017-02-21 17:48:00
Data0	ES	1129	2017-02-21 11:48:00	2359.75
Data1	SPY	2155	2017-02-16 16:00:00	234.72
Data2	ES-daily	2170	2017-02-16 15:15:00	2345.5
```

backfill source is configured as follows:
```python
# Parse static SPY data file
bfdata1 = btfeed.GenericCSVData(dataname=bfdata1path,
                                dtformat=('%Y-%m-%d'),
                                datetime=0,
                                high=1,
                                low=2,
                                open=3,
                                close=4,
                                volume=5,
                                openinterest=6,
                                sessionstart=dt.time(9, 30),
                                sessionend=dt.time(16, 0),
                                tz=EST)
```
Live data source is configured as follows:
```python
# SPY Live data timeframe resampled to 1 Day
data1 = ibstore.getdata(dataname=args.live_spy, backfill_from=bfdata1,
                        timeframe=bt.TimeFrame.Days, compression=1,
                        sessionstart=dt.time(9, 30),
                        sessionend=dt.time(16, 0),
                        tz='EST5EDT')
cerebro.resampledata(data1, name='SPY', timeframe=bt.TimeFrame.Days, compression=1)
data1.plotinfo.plot = False
```

**Comment by rterbush** [comment]
System startup worked properly yesterday 2017-02-22, but today, it does not show yesterday's closing bar for the backfilled daily data.

```
-- 5248 2017-02-23 14:20:00
Data0	ES	0921	2017-02-23 08:20:00	2366.25
Data1	SPY	2156	2017-02-21 16:00:00	236.49
Data2	ES-daily	2171	2017-02-21 15:15:00	2360.0
*** 2017-02-23T07:20:18 DATA NOTIF: ES LIVE
-- 5249 2017-02-23 14:21:00
Data0	ES	0922	2017-02-23 08:21:00	2366.25
Data1	SPY	2156	2017-02-21 16:00:00	236.49
Data2	ES-daily	2171	2017-02-21 15:15:00	2360.0
```

**Comment by rterbush** [comment]
After weekend gap and successfully displaying latest daily bars on Friday startup on 2017/2/24, not presenting the bar from Friday's close today.

```
-- 5273 2017-02-27 14:43:00
Data0	ES	0944	2017-02-27 08:43:00	2361.5
Data1	SPY	2157	2017-02-23 16:00:00	236.44
Data2	ES-daily	2172	2017-02-23 15:15:00	2362.75
-- 5274 2017-02-27 14:44:00
Data0	ES	0945	2017-02-27 08:44:00	2361.5
Data1	SPY	2157	2017-02-23 16:00:00	236.44
Data2	ES-daily	2172	2017-02-23 15:15:00	2362.75
*** 2017-02-27T07:44:32 DATA NOTIF: ES LIVE
-- 5275 2017-02-27 14:45:00
Data0	ES	0946	2017-02-27 08:45:00	2362.0
Data1	SPY	2157	2017-02-23 16:00:00	236.44
Data2	ES-daily	2172	2017-02-23 15:15:00	2362.75
```

**Comment by mementum** [comment]
The code to pull data from the `backfill_from` source is dead simple:
```python
                if not self.p.backfill_from.next():
                    # additional data source is consumed
                    self._state = self._ST_START
                    continue

                # copy lines of the same name
                for alias in self.lines.getlinealiases():
                    lsrc = getattr(self.p.backfill_from.lines, alias)
                    ldst = getattr(self.lines, alias)

                    ldst[0] = lsrc[0]

                return True
```

It won't stop until the data source is exhausted. The options to consider:

  1. The `backfill_from` has only data to the point shown. This should not be the case because you have checked the actual content.

  2. The resampling code is getting confused at some point and fails to deliver that last bar.

  3. The `datetime` interactions are confusing

And after checking the code it may be that number `3` is the winner. In `GenericCSVData` the following conversion takes place

```python
        self.lines.datetime[0] = date2num(dt)
```

This after using the `sessionend=dt.time(16, 0)` that you have specified. Because the high level `date2num` function is being used, rather than the method available in each data feed, `16:00` is being understood as `UTC` time.

There is another parameter to each data feed which is `tzinput`, to convert time if coming from already in a timezone specific format. Although it would actually help in this case it isn't the route to go.

Some of the original goals may have been too ambitious, simply by not forcing users to specify things like timeframes (and by python not having a time concept like `end of day`) A large thoughtful process has to take place.

**Comment by rterbush** [comment]
Reporting back on results of today's run with this code change.

If the pattern is valid, and I am actually seeing the missing bar every other day (which it seems may be the case), I should have NOT seen the bar for 2017/2/28 today. The good news is that the system started off and showed data for the daily backfills to include 2017/2/28.

However, (and not certain this is related), I did not get a print for the 16:00 closing bar on SPY today. I've not looked into this any deeper, but no code changed here between yesterday's close which I did see a closing bar, and today's close other than to update to development branch code.
